rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java

```diff
@@ -506,172 +510,150 @@ public void onFailure(Throwable caught) {
-
-            switch (nodeSource.getNodeSourceStatus()) {
-                case NODES_DEPLOYED:
-                    numDeployedNodeSources++;
-                    break;
-                case NODES_UNDEPLOYED:
-                    numUndeployedNodeSources++;
-                    break;
-            }
+        HashMap<String, NodeSource> newNodeSources = new HashMap<>();
+        if (isRegularRequest(currentCounter, latestCounter)) {
+            copyNodesSources(model.getNodeSources(), newNodeSources);
         }
 
-        model.setNumDeployedNodeSources(numDeployedNodeSources);
-        model.setNumUndeployedNodeSources(numUndeployedNodeSources);
+        processNodeSources(newNodeSources, obj);
 
-        int numDeploying = 0;
-        int numLost = 0;
-        int numConfiguring = 0;
-        int numFree = 0;
-        int numBusy = 0;
-        int numDown = 0;
-        int numToBeRemoved = 0;
+        // process nodes
+        JSONArray jsNodes = obj.get("nodesEvents").isArray();
+        for (int i = 0; i < jsNodes.size(); i++) {
+            try {
+                JSONObject jsNode = jsNodes.get(i).isObject();
 
-        int numLocked = 0;
+                final Node node = parseNode(jsNode);
 
-        JSONArray nodes = obj.get("nodesEvents").isArray();
-        for (int i = 0; i < nodes.size(); i++) {
-            try {
-                JSONObject nodeObj = nodes.get(i).isObject();
-
-                String hostName = nodeObj.get("hostName").isString().stringValue();
-                String nss = nodeObj.get("nodeSource").isString().stringValue();
-
-                String nodeUrl = nodeObj.get("nodeUrl").isString().stringValue();
-                String nodeState = nodeObj.get("nodeState").isString().stringValue();
-                String nodeInfo = nodeObj.get("nodeInfo").isString().stringValue();
-                String timeStampFormatted = nodeObj.get("timeStampFormatted").isString().stringValue();
-                long timeStamp = Math.round(nodeObj.get("timeStamp").isNumber().doubleValue());
-                String nodeProvider = nodeObj.get("nodeProvider").isString().stringValue();
-
-                String nodeOwner = getJsonStringNullable(nodeObj, "nodeOwner");
-                String vmName = getJsonStringNullable(nodeObj, "vmname");
-                String description = getJsonStringNullable(nodeObj, "nodeInfo");
-                String defaultJMXUrl = getJsonStringNullable(nodeObj, "defaultJMXUrl");
-                String proactiveJMXUrl = getJsonStringNullable(nodeObj, "proactiveJMXUrl");
-
-                boolean isLocked = getJsonBooleanNullable(nodeObj, "locked", false);
-                long lockTime = getJsonLongNullable(nodeObj, "lockTime", -1);
-                String nodeLocker = getJsonStringNullable(nodeObj, "nodeLocker");
-
-                if (isLocked) {
-                    numLocked++;
-                }
+                final NodeSource nodeSource = newNodeSources.get(node.getSourceName());
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Add a default case to this switch. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this "switch" statement by "if" statements to increase readability. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)